### PR TITLE
CONFIG: update versions for paraview

### PR DESCRIPTION
**Pull request by olesenm**
- drop old TCL support from paraview build.

- add +plugins variant to have include directories installed. This is
  enabled by default since the additional diskspace for includes is
  really minimal and since this also allows re-use of the VTK libraries
  from ParaView without necessarily requiring a separate VTK
  installation.

### CODE REVIEW

**Comment by olesenm** [comment]
OK, let's rephrase it to 'now also installs its header files'.
The point about re-use elsewhere is simply that VTK libraries and headers are used by both vtk and paraview. If I write an application that needs some VTK components, the cmake can explore and use the vtk lib/include from either a 'plain' VTK or ParaView. Without the includes from ParaView, it's impossible to write any extensions or plugins for it (or to use it's libraries anywhere else in that case).

**Comment by adamjstewart** [comment]
Can you fix the flake8 warnings?

**Comment by davydden** [comment]
that's what flake8 complains about:
```
var/spack/repos/builtin/packages/paraview/package.py:48: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/paraview/package.py:55: [E265] block comment should start with '# '
var/spack/repos/builtin/packages/paraview/package.py:56: [E265] block comment should start with '# '
var/spack/repos/builtin/packages/paraview/package.py:81: [E303] too many blank lines (2)
var/spack/repos/builtin/packages/paraview/package.py:114: [E303] too many blank lines (2)
var/spack/repos/builtin/packages/paraview/package.py:114: [E501] line too long (80 > 79 characters)
```

**Comment by adamjstewart** [comment]
I always wonder, can everyone see the "Details" link next to Travis and view the output or only some people?

**Comment by olesenm** [comment]
Adjusted to make flake8 happy, removed some crufty looking code.

File: var/spack/repos/builtin/packages/paraview/package.py [hunk 1]
```diff
@@ -27,28 +27,31 @@
 
 class Paraview(Package):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/paraview/package.py]: One last request, can you convert this to `CMakePackage`?

**Comment by olesenm** [comment]
@adamjstewart @davydden please recheck.

File: var/spack/repos/builtin/packages/paraview/package.py (hunk 1 above)
Comment (olesenm) [inline var/spack/repos/builtin/packages/paraview/package.py]: Sorry, I did try that, but everything seemed to fall apart on me. I think that the cmake_args were getting stepped on somewhere. At least the reworked version is _really_ close to being a CMakePackage. But I'd really like to get this merged ASAP so that I can followup on my OpenFOAM packages (3 of them now), which depend on getting the includes from this updated paraview package.

Comment (olesenm) [inline var/spack/repos/builtin/packages/paraview/package.py]: OK, got it. The problem isn't CMakePackage in itself. (ie, spack install paraview is OK). The problem happens when I try `spack build paraview`. Then it goes crazy and bombs out while resolving hwloc for openmpi.

I'll make the Cmake change

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/paraview/package.py]: That's strange, how exactly is it crashing? I wonder if `spack build` tries to only run up to the build stage for every dependency.

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/paraview/package.py]: Nope, that's not the case. I would be interested to here what problems you are seeing.

**Comment by olesenm** [comment]
@adamjstewart The conversation disappeared, but here are the messages in all their glory.
With `spack install paraview` - no issues. Purge everything and run `spack build paraview`, then this is the traceback:

    spack build paraview @5.2.0
    ==> Checking dependencies for ['paraview', '@5.2.0']
    ==> bzip2 is already installed in /home/mol/Software/spack/opt/spack/linux-opensuse42-x86_64/gcc-4.8/bzip2-1.0.6-4ad5ytasjvlcvn3jg5vsubvwb6gjmu6n
    ==> zlib is already installed in /home/mol/Software/spack/opt/spack/linux-opensuse42-x86_64/gcc-4.8/zlib-1.2.11-gtj3fhsovk5gnnowdqdwvifnlhtnmuya
    ==> cmake is already installed in /home/mol/Software/spack/opt/spack/linux-opensuse42-x86_64/gcc-4.8/cmake-3.7.2-rnjyonjhambdkoromzzck6xrbnkqhnb3
    ==> qt is already installed in /home/mol/Software/spack/opt/spack/linux-opensuse42-x86_64/gcc-4.8/qt-4.8.6-mc5zfxk7opfeo3brbpc7qcq5ng4vbkue
    ==> libxml2 is already installed in /home/mol/Software/spack/opt/spack/linux-opensuse42-x86_64/gcc-4.8/libxml2-2.9.4-2rdcuf44tjbtacu5i5f66iqooc2aid3a
    ==> libtiff is already installed in /home/mol/Software/spack/opt/spack/linux-opensuse42-x86_64/gcc-4.8/libtiff-4.0.6-ft25ziwhcy3m5pfahbv7m3qat7cffeht
    ==> freetype is already installed in /home/mol/Software/spack/opt/spack/linux-opensuse42-x86_64/gcc-4.8/freetype-2.7-xtr77y6vge4dwiiibmfxepy5r34xhrh5
    ==> jpeg is already installed in /home/mol/Software/spack/opt/spack/linux-opensuse42-x86_64/gcc-4.8/jpeg-9b-76m6ni25j4jnk75w2gmmvb7ltynxgsdt
    ==> Installing openmpi
    ==> Using cached archive: /home/mol/Software/spack/var/spack/cache/openmpi/openmpi-2.1.0.tar.bz2
    ==> Staging archive: /home/mol/Software/spack/var/spack/stage/openmpi-2.1.0-stq373j5o2tdhyudyhf6a6w6tfcdmpfa/openmpi-2.1.0.tar.bz2
    ==> Created stage in /home/mol/Software/spack/var/spack/stage/openmpi-2.1.0-stq373j5o2tdhyudyhf6a6w6tfcdmpfa
    ==> Ran patch() for openmpi
    ==> Building openmpi [AutotoolsPackage]
    ==> Executing phase : 'autoreconf'
    ==> Executing phase : 'configure'
    ==> Error: KeyError: 'No spec with name hwloc in openmpi@2.1.0%gcc@4.8~java~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm~verbs+vt arch=linux-opensuse42-x86_64'
    /home/mol/Software/spack/var/spack/repos/builtin/packages/openmpi/package.py:223, in configure_args:
      176      def configure_args(self):
      177          spec = self.spec
      178
      179          config_args = [
      180              '--enable-shared',
      181              '--enable-static',
      182              '--enable-mpi-cxx',
      183              # Schedulers
      184              '--with-tm' if '+tm' in spec else '--without-tm',
      185              '--with-slurm' if '+slurm' in spec else '--without-slurm',
      186              # Fabrics
      187              '--with-psm' if '+psm' in spec else '--without-psm',
      188          ]
      189
      190          # Intel PSM2 support
      191          if spec.satisfies('@1.10:'):
      192              if '+psm2' in spec:
      193                  config_args.append('--with-psm2')
      194              else:
      195                  config_args.append('--without-psm2')
      196
      197          # PMI support
      198          if spec.satisfies('@1.5.5:'):
      199              if '+pmi' in spec:
      200                  config_args.append('--with-pmi')
      201              else:
      202                  config_args.append('--without-pmi')
      203
      204          # Mellanox Messaging support
      205          if spec.satisfies('@1.5.4:'):
      206              if '+mxm' in spec:
      207                  config_args.append('--with-mxm')
      208              else:
      209                  config_args.append('--without-mxm')
      210
      211          # OpenFabrics verbs support
      212          if '+verbs' in spec:
      213              path = _verbs_dir()
      214              if path is not None and path not in ('/usr', '/usr/local'):
      215                  config_args.append('--with-{0}={1}'.format(self.verbs, path))
      216              else:
      217                  config_args.append('--with-{0}'.format(self.verbs))
      218          else:
      219              config_args.append('--without-{0}'.format(self.verbs))
      220
      221          # Hwloc support
      222          if spec.satisfies('@1.5.2:'):
    >>223              config_args.append('--with-hwloc={0}'.format(spec['hwloc'].prefix))
      224
      225          # Java support
      226          if spec.satisfies('@1.7.4:'):
      227              if '+java' in spec:
      228                  config_args.extend([
      229                      '--enable-java',
      230                      '--enable-mpi-java',
      231                      '--with-jdk-dir={0}'.format(spec['jdk'].prefix)
      232                  ])
      233              else:
      234                  config_args.extend([
      235                      '--disable-java',
      236                      '--disable-mpi-java'
      237                  ])
      238
      239          # SQLite3 support
      240          if spec.satisfies('@1.7.3:1.999'):
      241              if '+sqlite3' in spec:
      242                  config_args.append('--with-sqlite3')
      243              else:
      244                  config_args.append('--without-sqlite3')
      245
      246          # VampirTrace support
      247          if spec.satisfies('@1.3:1.999'):
      248              if '+vt' not in spec:
      249                  config_args.append('--enable-contrib-no-build=vt')
      250
      251          # Multithreading support
      252          if spec.satisfies('@1.5.4:'):
      253              if '+thread_multiple' in spec:
      254                  config_args.append('--enable-mpi-thread-multiple')
      255              else:
      256                  config_args.append('--disable-mpi-thread-multiple')
      257
      258          return config_args

**Comment by adamjstewart** [comment]
@alalazo Can you take a look at that last error message? I can't imagine why `spack build` and `spack install` would have different specs in the dependency chain.

File: var/spack/repos/builtin/packages/paraview/package.py [hunk 2]
```diff
@@ -25,30 +25,36 @@
...
     depends_on('python@2:2.7', when='+python')
     depends_on('py-numpy', when='+python', type='run')
     depends_on('py-matplotlib', when='+python', type='run')
-    depends_on('tcl', when='+tcl')
     depends_on('mpi', when='+mpi')
     depends_on('qt@:4', when='+qt')
+    # TODO# depends_on('qt@:4', when='@:5.2.0+qt')
+    # TODO# depends_on('qt@5',  when='@5.3.0:+qt')
 
     depends_on('cmake', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/paraview/package.py]: You can remove this line now. It's already in `CMakePackage`.

File: var/spack/repos/builtin/packages/paraview/package.py [hunk 3]
```diff
@@ -65,61 +71,67 @@ class Paraview(Package):
...
+            return off
+
+        def nvariant_bool(feature):
+            """Negated ternary for spec variant to OFF/ON string"""
+            return variant_bool(feature, on='OFF', off='ON')
+
+        rendering = variant_bool('+opengl2', 'OpenGL2', 'OpenGL')
+        includes  = variant_bool('+plugins')
+
+        cmake_args = std_cmake_args[:]
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/paraview/package.py]: You don't need this line anymore. `std_cmake_args` are already added the the `cmake` call in `CMakePackage`. See `spack edit -b cmake` for details.

**Comment by alalazo** [comment]
@adamjstewart 

> Can you take a look at that last error message? I can't imagine why spack build and spack install would have different specs in the dependency chain.

I'll try to have a look at it tomorrow.

File: var/spack/repos/builtin/packages/paraview/package.py (hunk 2 above)
Comment (olesenm) [inline var/spack/repos/builtin/packages/paraview/package.py]: done

File: var/spack/repos/builtin/packages/paraview/package.py (hunk 3 above)
Comment (olesenm) [inline var/spack/repos/builtin/packages/paraview/package.py]: done

**Comment by olesenm** [comment]
OK - apart from the previously mentioned oddness with `build` vs. `install` I think that I have everything in that is directly related to paraview and satisfied all the requested changes.
Could someone please merge this so that I can get my pending openfoam merge request addressed?

**Comment by alalazo** [comment]
The problem above doesn't seem related to `spack build`. I could reproduce the same issue with:
```console
$ spack install --only=dependencies paraview
```
on `LLNL/develop`.

**Comment by alalazo** [comment]
FYI I opened #3581 to keep track of the issue
